### DISCUSSION TITLE
Question about type_error::create

### DISCUSSION THREAD

**Discussion by jamesbloom-scheff** [nlohmann/json#3566] [Q&A] [answered]
I'm looking at the following piece of code

```
        throw nlohmann::json::type_error::create(
            302,
            "type must be object, but is " + std::string(json.type_name()));
```

You will notice that create has two arguments. All the documentation I see online shows only two arguments

But the definition in exceptiobs.hpp seems to show three arguments

```
class type_error : public exception
{
  public:
    template<typename BasicJsonContext, enable_if_t<is_basic_json_context<BasicJsonContext>::value, int> = 0>
    static type_error create(int id_, const std::string& what_arg, BasicJsonContext context)
    {
        std::string w = concat(exception::name("type_error", id_), exception::diagnostics(context), what_arg);
        return {id_, w.c_str()};
    }
```

Visual studio 2019 generates an error
error C2780: 'nlohmann::detail::type_error nlohmann::detail::type_error::create(int,const std::string &,BasicJsonContext)': expects 3 arguments - 2 provided

Can someone explain what is going on and how to fix it?

I thought I was using 3.10.5 but the documentation seems to say 3.10.2
The software I am looking at is documented as using 3.0.2

**Comment by falbrechtskirchinger** [comment, marked as answer]
What exactly are you trying to do? Please provide more context.

> You will notice that create has two arguments. All the documentation I see online shows only two arguments

Where are you seeing this? Creating exceptions is not documented at all, to my knowledge.

> I thought I was using 3.10.5 but the documentation seems to say 3.10.2
> ```cpp
> template<typename BasicJsonContext, enable_if_t<is_basic_json_context<BasicJsonContext>::value, int> = 0>
>     static type_error create(int id_, const std::string& what_arg, BasicJsonContext context)
> ```

Based on this code, you're not using 3.10.5 but `develop`. In 3.10.5 the third parameter is of type `const BasicJsonType &`.

> Visual studio 2019 generates an error
> error C2780: 'nlohmann::detail::type_error nlohmann::detail::type_error::create(int,const std::string &,BasicJsonContext)': expects 3 arguments - 2 provided

Where in the library is this error occurring? If it's not library code that is throwing this `type_error`, you're out of luck as far as support is concerned. But if you don't care about your code breaking (again) in a future release, you can just pass `nullptr`.

**Reply by jamesbloom-scheff** [reply]
I am trying to modify code to build and run under windows 10 and visual studio. Code that was written originally written years ago for linux only. In addition, it was written for 64 bits and I need to get it to work for 32 bits. 

Are you saying that I am using the wrong version of nlohmann json? That develop is different from 3.10.5?

Yes, I didn't find much documentation (or examples) on how to throw a type_error exception, but I found a few.

Are people not supposed to throw type_error exceptions? If they are, how many arguments does it have? Has the number of arguments changed? Presumably, this code used to build.

Is that third argument supposed t default to 0?

**Reply by nlohmann** [reply]
You must not throw exceptions from the library. In which use case do you do that?

**Reply by falbrechtskirchinger** [reply]
> I am trying to modify code to build and run under windows 10 and visual studio. Code that was written originally written years ago for linux only. In addition, it was written for 64 bits and I need to get it to work for 32 bits.
> 
> Are you saying that I am using the wrong version of nlohmann json? That develop is different from 3.10.5?

Yes. The `develop` branch, as the name suggests, contains the in-development version of the library. 

> Yes, I didn't find much documentation (or examples) on how to throw a type_error exception, but I found a few.

The official documentation can be found here: https://json.nlohmann.me/

> Are people not supposed to throw type_error exceptions?

No. The exceptions are not part of the public API, in the sense that users are not supposed to create or throw them.

> If they are, how many arguments does it have? Has the number of arguments changed? Presumably, this code used to build.

I started contributing after the 3.10.5 release. I don't know the history of the library that well. All I can say is that it's been 3 arguments since at least 3.10.5 and that I changed the type of the 3rd argument. That change will be part of the 3.11.0 release.

> Is that third argument supposed t default to 0?

That depends on the version. In 3.10.5 it's supposed to be an instance of `basic_json`, later it changed to a pointer of `basic_json` or `nullptr`. It's a bad idea to depend on that though. It can change without warning.

**Comment by jamesbloom-scheff** [comment]
So, It seems like my problem is that the original code that I am trying to modify tried to throw an exception of type "type_error" which is unsupported. Should user code be trying to throw any type of nlohmann::json exception? I suspect not.

The JSON.hpp has the following:
 __|  |   __|     |   | |  JSON for Modern C++
|  |  |__   |  |  | | | |  version 3.10.5
|_____|_____|_____|_|___|  https://github.com/nlohmann/json

Licensed under the MIT License <http://opensource.org/licenses/MIT>.
SPDX-License-Identifier: MIT
Copyright (c) 2013-2022 Niels Lohmann <http://nlohmann.me>.

Are you sure that I am using develop?

**Reply by gregmarr** [reply]
>Are you sure that I am using develop?

Here is the commit where it changed to what you showed: https://github.com/nlohmann/json/pull/3415/files#diff-512022761059423c28c91c9843d15295e0b030c23f0024af68ea4ecac8823d07R139

That is after the 3.10.5 release, and a new release hasn't been made yet, so the json.hpp in develop still has the 3.10.5 version number listed.

**Reply by falbrechtskirchinger** [reply]
> So, It seems like my problem is that the original code that I am trying to modify tried to throw an exception of type "type_error" which is unsupported. Should user code be trying to throw any type of nlohmann::json exception? I suspect not.

As Niels commented [here](https://github.com/nlohmann/json/discussions/3566#discussioncomment-3064186) - no.

> The JSON.hpp has the following: **| | **| | | | JSON for Modern C++ | | |** | | | | | | version 3.10.5 |**_**|**_**|**_**|_|_**| https://github.com/nlohmann/json
> 
> Licensed under the MIT License http://opensource.org/licenses/MIT. SPDX-License-Identifier: MIT Copyright (c) 2013-2022 Niels Lohmann http://nlohmann.me.
> 
> Are you sure that I am using develop?

The version number in the header is updated as part of the release process. Development versions simply carry the version information from the last release.

**Reply by nlohmann** [reply]
@jamesbloom-scheff Do you need further assistance?
